File: bot/src/command/GrantCommand.ts
import Command from "./Command";
import {VkMessage} from "../service/VkMessagesService";

export default class GrantCommand extends Command {
    canYouHandleThisCommand(command: string, message: VkMessage): boolean {
        return command == "grant";
    }

    getCommandShortUsage(): string {
        return "/sweet grant (plan)[:days] (user)";
    }

    requiresPrivileges(peerId: number): boolean {
        return true;
    }

    async handle(command: string, rawArguments: string, message: VkMessage): Promise<void> {
        const { vkMessagesService, usagePlanOrmService } = this.context;

        const firstArg = rawArguments.split(" ")[0];
        let planId: string | null;
        let days: number | null = null;
        if (firstArg.includes(":")) {
            const parts = firstArg.split(":");
            planId = parts[0];
            days = +parts[1];
        } else {
            planId = firstArg;
        }
        if (!planId) {
            await vkMessagesService.send(message.peerId, this.getUsage());
            return;
        }
        const plan = await usagePlanOrmService.getUsagePlan(planId);
        if (!plan) {
            await vkMessagesService.send(message.peerId, 'Нет такого плана');
            return;
        }
        const restArgs = rawArguments.substring(firstArg.length + 1);
        let memberId: number | null;
        if (restArgs.length > 0) {
            memberId = this.extractMemberId(restArgs);
            if (memberId == null) {
                await vkMessagesService.send(message.peerId, `Не могу найти такого пользователя`);
                return;
            }
        } else if (message.forwardedMessages.length == 1) {
            memberId = message.forwardedMessages[0].fromId;
        } else {
            await vkMessagesService.send(message.peerId, this.getUsage());
            return;
        }
        const user = await this.context.vkUsersService.getUser(memberId);
        if (!user) {
            await vkMessagesService.send(message.peerId, `Не могу найти такого пользователя`);
            return;
        }
        const expiry = days ? new Date(Date.now() + days * 24 * 60 * 60 * 1000) : null;
        await this.context.vkUsersOrmService.setUserUsagePlan(memberId, planId, expiry);
        await vkMessagesService.send(message.peerId, `✨ [id${Math.abs(memberId)}|Ты] получаешь Sweetie AI ${plan.title} ${expiry ? `до ${expiry.getDate()}.${expiry.getMonth() + 1}.${expiry.getFullYear()}` : 'навсегда'}`);
    }

    private extractMemberId(text: string): number | null {
        const regex = /\[?(id|club)(\d+)\|?/;
        const match = text.match(regex);
        if (match && !isNaN(+match[2])) {
            return +match[2];
        } else {
            return null;
        }
    }

    private getUsage(): string {
        let result = '';
        result += 'Так пиши:\n';
        result += '/sweet grant (plan)[:days] [user]\n';
        return result;
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/command/GrantCommand.kt
package space.davids_digital.sweetie.command

import jakarta.transaction.Transactional
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.integration.vk.VkMessageService
import space.davids_digital.sweetie.integration.vk.VkUtils
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.orm.repository.UsagePlanRepository
import space.davids_digital.sweetie.orm.repository.VkUserRepository
import java.time.ZonedDateTime

@Component
class GrantCommand(
    private val vkMessageService: VkMessageService,
    private val usagePlanRepository: UsagePlanRepository,
    private val vkUserRepository: VkUserRepository
): Command {
    companion object {
        private val log = LoggerFactory.getLogger(GrantCommand::class.java)
    }

    override fun getNames(): Array<String> {
        return arrayOf("grant")
    }

    override fun getUsage(): String {
        return "grant (plan)[:days] (user)"
    }

    override fun requiresChatAdmin(): Boolean {
        return true
    }

    override fun requiresAppCeo(): Boolean {
        return true
    }

    @Transactional
    override suspend fun handle(commandName: String, rawArguments: String, message: VkMessageModel) {
        if (rawArguments.isBlank()) {
            return handleHelp(message)
        }
        val firstArgument = rawArguments.split(" ")[0]
        val planId = firstArgument.split(":")[0]
        val days = firstArgument.split(":").getOrNull(1)?.toIntOrNull()
        val rest = rawArguments.drop(firstArgument.length).trim()

        var memberId = VkUtils.extractMemberId(rest)
        if (memberId == null) {
            memberId = message.forwardedMessages.firstOrNull()?.fromId
        }
        if (memberId == null) {
            return handleHelp(message)
        }

        val planExists = withContext(Dispatchers.IO) {
            usagePlanRepository.existsById(planId)
        }
        if (!planExists) {
            vkMessageService.send(message.peerId, "Нет такого плана")
            return
        }

        val expiry = days?.let { ZonedDateTime.now().plusDays(it.toLong()) }
        withContext(Dispatchers.IO) {
            vkUserRepository.setUsagePlan(memberId, planId, expiry)
        }
        log.info("User ${message.fromId} set plan '$planId' to $memberId ${days?.let { "for $it days" } ?: "forever"}")

        val userTag = if (memberId > 0) "[id$memberId|Ты]" else "Ты"
        val expiryText = expiry?.let { "до ${it.dayOfMonth}.${it.monthValue}.${it.year}" } ?: "навсегда"
        val planTitle = withContext(Dispatchers.IO) {
            usagePlanRepository.findById(planId).get().title
        }
        vkMessageService.send(message.peerId, "$userTag получаешь Sweetie AI $planTitle $expiryText")
    }

    private fun handleHelp(message: VkMessageModel) {
        val text = """
            Так пиши:
            /sweet grant (plan)[:days] [user]
        """.trimIndent()
        vkMessageService.send(message.peerId, text)
    }
}